61/api pagination is not reflected in the app

Odonata971 opened the pull request:
This pull request introduces significant changes to the `Home` activity and related classes in the `pathpilot_android_client` project, focusing on implementing pagination for clients and itineraries. The most important changes include adding new `ClientPage` and `ItineraryPage` classes, updating the `Home` activity to handle these pages, and modifying the `ClientService` and `ItineraryService` to support fetching paginated data.

### Pagination Implementation:

* [`app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/clients/ClientPage.java`](diffhunk://#diff-2ceca4a91da666f1e722106cc706f4238d543603e3d087a53f797665e0a1deedR1-R26): Added a new `ClientPage` class to represent a page of clients with links to other pages.
* [`app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/itinerary/ItineraryPage.java`](diffhunk://#diff-992896875d5f268968b5e02c604bf50f25dca9eee71e44855b571649ff3fa531R1-R26): Added a new `ItineraryPage` class to represent a page of itineraries with links to other pages.

### Home Activity Updates:

* [`app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/Home.java`](diffhunk://#diff-01ef66333cbcb71c1325283add7d81746e01a5dba6be761b8aacc50c6001fe39R27-R40): Updated the `Home` activity to use the new `ClientPage` and `ItineraryPage` classes and handle their interactions. [[1]](diffhunk://#diff-01ef66333cbcb71c1325283add7d81746e01a5dba6be761b8aacc50c6001fe39R27-R40) [[2]](diffhunk://#diff-01ef66333cbcb71c1325283add7d81746e01a5dba6be761b8aacc50c6001fe39R54-R56) [[3]](diffhunk://#diff-01ef66333cbcb71c1325283add7d81746e01a5dba6be761b8aacc50c6001fe39R141-R169)

### Service Updates:

* [`app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/clients/ClientService.java`](diffhunk://#diff-609dd0bffbaa62ea929b72734274bb4d3dcadea6dac813e6777f2fbd1aec4b5aL20): Modified the `ClientService` to fetch paginated client data and save the current page to the activity. [[1]](diffhunk://#diff-609dd0bffbaa62ea929b72734274bb4d3dcadea6dac813e6777f2fbd1aec4b5aL20) [[2]](diffhunk://#diff-609dd0bffbaa62ea929b72734274bb4d3dcadea6dac813e6777f2fbd1aec4b5aL38-R37) [[3]](diffhunk://#diff-609dd0bffbaa62ea929b72734274bb4d3dcadea6dac813e6777f2fbd1aec4b5aL54-R118)
* [`app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/itinerary/ItineraryService.java`](diffhunk://#diff-50b3fd0e54f5e08a1cc32d3db4a462b2717edbc75daf71cf8a83c4e275261103L12-R12): Modified the `ItineraryService` to fetch paginated itinerary data and save the current page to the activity. [[1]](diffhunk://#diff-50b3fd0e54f5e08a1cc32d3db4a462b2717edbc75daf71cf8a83c4e275261103L12-R12) [[2]](diffhunk://#diff-50b3fd0e54f5e08a1cc32d3db4a462b2717edbc75daf71cf8a83c4e275261103L21-R34) [[3]](diffhunk://#diff-50b3fd0e54f5e08a1cc32d3db4a462b2717edbc75daf71cf8a83c4e275261103L102-R114)

### Fragment Updates:

* [`app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/clients/FragmentClients.java`](diffhunk://#diff-d45f185d13accf54003f1828c147129884cc9c8b48ea42d0b45ea1b5760b8060R3-R4): Updated the `FragmentClients` to load more clients when reaching the bottom of the list using the new pagination system. [[1]](diffhunk://#diff-d45f185d13accf54003f1828c147129884cc9c8b48ea42d0b45ea1b5760b8060R3-R4) [[2]](diffhunk://#diff-d45f185d13accf54003f1828c147129884cc9c8b48ea42d0b45ea1b5760b8060R15) [[3]](diffhunk://#diff-d45f185d13accf54003f1828c147129884cc9c8b48ea42d0b45ea1b5760b8060R28) [[4]](diffhunk://#diff-d45f185d13accf54003f1828c147129884cc9c8b48ea42d0b45ea1b5760b8060R50-R51) [[5]](diffhunk://#diff-d45f185d13accf54003f1828c147129884cc9c8b48ea42d0b45ea1b5760b8060R78-R96) [[6]](diffhunk://#diff-d45f185d13accf54003f1828c147129884cc9c8b48ea42d0b45ea1b5760b8060L108-R141)
* [`app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/itinerary/FragmentItineraries.java`](diffhunk://#diff-71c6196eb396eccecb560336652a309ac1875a3e56ed642cac0ae601be752e03R3-R4): Updated the `FragmentItineraries` to load more itineraries when reaching the bottom of the list using the new pagination system. [[1]](diffhunk://#diff-71c6196eb396eccecb560336652a309ac1875a3e56ed642cac0ae601be752e03R3-R4) [[2]](diffhunk://#diff-71c6196eb396eccecb560336652a309ac1875a3e56ed642cac0ae601be752e03R15) [[3]](diffhunk://#diff-71c6196eb396eccecb560336652a309ac1875a3e56ed642cac0ae601be752e03L24-R27) [[4]](diffhunk://#diff-71c6196eb396eccecb560336652a309ac1875a3e56ed642cac0ae601be752e03R75-R93) [[5]](diffhunk://#diff-71c6196eb396eccecb560336652a309ac1875a3e56ed642cac0ae601be752e03L116-R139)

Commit:
feat: add client pagination support

- Introduced ClientPage to manage clients and links.
- Updated ClientService to handle API responses with pagination.
- Implemented loading more clients on scroll in FragmentClients.
- Removed old pagination layout and drawable resources.
- Enhanced Parser to extract pagination links from API response.

Commit:
feat: Refactor itinerary handling and pagination

- Added correct pagination behaviors for itinerary list
- Added pagination support for itineraries in the UI.
- Implemented loading more itineraries on scroll.
- Enhanced parsing logic to accommodate new data structure. → extracted method to parse pagination links whicj are not related to the client or itineraries rather than the page itself

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/clients/ClientService.java:
```diff
@@ -51,19 +50,72 @@ public static void getClients(Context context, ListView listClientsView) {
 
-        // TODO Update URL to use the pagination
-        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, API_BASE_URL + "/all", null,
+        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, API_BASE_URL, null,
```
Use `createAuthenticatedRequest` in [NetworkUtils](https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/blob/90ac28d8280bf9e6baa305767799d6ce0fe23b67/app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/network/NetworkUtils.java)
https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/blob/90ac28d8280bf9e6baa305767799d6ce0fe23b67/app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/network/NetworkUtils.java#L105

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/clients/ClientService.java:
```diff
@@ -51,19 +50,72 @@ public static void getClients(Context context, ListView listClientsView) {
+        ProgressDialog progressDialog = new ProgressDialog(context);
+        progressDialog.show();
+
+        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, nextPageUrl, null,
```
Same use `createAuthenticatedRequest`

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/home/itinerary/ItineraryService.java:
```diff
@@ -124,6 +130,57 @@ public Map<String, String> getHeaders() {
+        ProgressDialog progressDialog = new ProgressDialog(context);
+        progressDialog.show();
+
+        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, nextPageUrl, null,
```
Same use `createAuthenticatedRequest`

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/Login.java:
```diff
@@ -54,6 +54,10 @@ protected void onCreate(Bundle savedInstanceState) {
             emailInput.setText(intent.getStringExtra(CLE_MAIL));
         }
+
+        // TODO remove this
```
Remove THAT

Francois389 commented on app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Link.java:
```diff
@@ -0,0 +1,9 @@
+/**
+ * Represents a link in a REST API response
+ */
+public record Link(@NonNull String rel, @NonNull String href) {
```
Can you explain what is a `rel`

Commit:
feat: simplify network requests with auth handling

- Replaced manual header management in network requests with a utility function for authenticated requests.
- Updated client and itinerary services to use the new request method.
- Removed unnecessary imports and commented-out code from login functionality.
- Enhanced documentation for the Link class to clarify its purpose.

Francois389 reviewed (approved):
LGTM
